File: Week_02/kotlin/src/_103_BinaryTreeZigzagLevelOrderTraversal.kt
import java.util.*

class _103_BinaryTreeZigzagLevelOrderTraversal_ {
    /**
     * Example:
     * var ti = TreeNode(5)
     * var v = ti.`val`
     * Definition for a binary tree node.
     * class TreeNode(var `val`: Int) {
     *     var left: TreeNode? = null
     *     var right: TreeNode? = null
     * }
     */
    class Solution {
        fun zigzagLevelOrder(root: TreeNode?): List<List<Int>> {
            root ?: return emptyList()
            val q = LinkedList<TreeNode>().apply { offer(root) }
            val ret = mutableListOf<List<Int>>()
            var isOrderLeft = true
            while (q.isNotEmpty()) {
                val deq = LinkedList<Int>()
                var size = q.size
                while (size-- > 0) {
                    val node = q.poll()
                    if (isOrderLeft) {
                        deq.offer(node.`val`)
                    } else {
                        deq.offerFirst(node.`val`)
                    }
                    node.left?.let { q.offer(it) }
                    node.right?.let { q.offer(it) }
                }
                isOrderLeft = !isOrderLeft
                ret += deq
            }
            return ret
        }
    }
}

File: Week_02/cpp/src/_103_BinaryTreeZigzagLevelOrderTraversal.cpp
#include <stack>
#include <queue>
#include <deque>
#include <vector>
#include <unistd.h>

using namespace std;
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
};
class MySolution
{
public:
    // O(n) O(n)
    // 按照节点遍历顺序打印值
    vector<vector<int>> zigzagLevelOrder(TreeNode *root)
    {
        if (!root)
            return {};
        vector<vector<int>> ret;
        stack<TreeNode *> s[2];
        int idx = 0;
        s[idx].push(root);
        while (!s[idx].empty())
        {
            auto &s1 = s[idx], &s2 = s[idx ^ 1];
            vector<int> level;
            while (!s1.empty())
            {
                auto node = s1.top();
                s1.pop();
                level.push_back(node->val);
                if (!idx)
                {
                    if (node->left)
                        s2.push(node->left);
                    if (node->right)
                        s2.push(node->right);
                }
                else
                {
                    if (node->right)
                        s2.push(node->right);
                    if (node->left)
                        s2.push(node->left);
                }
            }
            idx ^= 1;
            ret.push_back(level);
        }
        return ret;
    }
};
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution
{
public:
    // O(n) O(n)
    // 按照节点遍历顺序逆序打印值
    vector<vector<int>> zigzagLevelOrder(TreeNode *root)
    {
        if (!root)
            return {};
        vector<vector<int>> ret;
        bool isOrderLeft = true;
        queue<TreeNode *> q;
        q.push(root);
        while (q.size())
        {
            auto size = q.size();
            deque<int> level;
            while (size-- > 0)
            {
                auto node = q.front();
                q.pop();
                if (isOrderLeft)
                {
                    level.push_back(node->val);
                }
                else
                {
                    level.push_front(node->val);
                }
                if (node->left)
                    q.push(node->left);
                if (node->right)
                    q.push(node->right);
            }
            ret.push_back(vector<int>{level.begin(), level.end()});
            isOrderLeft = !isOrderLeft;
        }
        return ret;
    }
};
int main(int argc, char const *argv[])
{
    char *msg = "hi\n";
    write(1, msg, strlen(msg));
    return 0;
}
